src/core/declarativewebcontainer.cpp

```diff
@@ -851,17 +852,17 @@ void DeclarativeWebContainer::updateLoading()
     emit loadingChanged();
 }
 
-void DeclarativeWebContainer::onFirstPaint(int, int)
-{
-    updateActiveTabRendered();
-}
-
 void DeclarativeWebContainer::updateActiveTabRendered()
 {
+    if (m_webPage && !m_webPage->completed()) {
```

Comment:
Gecko cannot render the active web page before we set the page as active [1]. In other words, during time period between creation start and finish the newly created web page cannot render. Browser takes care of tracking active web page.

Rendering sequence [2] goes as follows, in case you're interested:

When widget is valid (our EmbedLitePuppetWidget that is hosting window)
- First let window to draw underlay (e.i. we're doing custom clearing)
- Do rendering
- Draw custom overlay
- Do post rendering steps (ends up emitting compositingFinished)

The EmbedLitePupperWidget (nsIWidget) implements above methods [3] and forwards calls to EmbedLiteWindowListener which in turn is implemented by QMozWindow private. All of this is done in gecko compositor thread. Due to that in this case we don't use direct connection but unique connection (takes case switching tabs) so that all is done on the main thread.

[1]
https://git.merproject.org/mer-core/qtmozembed/blob/nemo_embedlite_38/src/qopenglwebpage.cpp#L114

[2]
https://git.merproject.org/mer-core/gecko-dev/blob/nemo_embedlite_38/gfx/layers/composite/LayerManagerComposite.cpp#L724

[3]
https://git.merproject.org/mer-core/gecko-dev/blob/nemo_embedlite_38/embedding/embedlite/embedshared/EmbedLitePuppetWidget.cpp#L755